File: 프로그래머스/lv2/12902. 3 x n 타일링/3 x n 타일링.py
def solution(n):
    answer = 0
    dp = [0]*(n+1)
    dp[0] = 1
    dp[2] = 3
    
    if n%2 != 0:
        return 0
    for i in range(4, n+1, 2):
        dp[i] += 3*dp[i-2]%1000000007
        for j in range(0, i-4+1, 2):
            dp[i] += 2*dp[j]%1000000007
        
    return dp[n]%1000000007

File: 프로그래머스/2/12902. 3 x n 타일링/3 x n 타일링.java
class Solution {
    public int solution(int n) {
        if (n % 2 != 0) {
            return 0;  
        }

        int MOD = 1000000007;
        long[] dp = new long[n + 1];

        dp[0] = 1;  
        if (n >= 2) {
            dp[2] = 3;  
        }

        for (int i = 4; i <= n; i += 2) {
            dp[i] = (dp[i] + 3 * dp[i - 2]) % MOD;

            for (int j = 0; j <= i - 4; j += 2) {
                dp[i] = (dp[i] + 2 * dp[j]) % MOD;
            }
        }

        return (int) dp[n];
    }
}
